whylog/constraints/__init__.py

```diff
@@ -2,42 +2,189 @@
+        self._check_constructor_groups()
+        self._check_constructor_params()
+
+    def _check_constructor_groups(self):
+        groups_count = len(self.groups)
+        if (self.MIN_GROUPS_COUNT is not None and groups_count < self.MIN_GROUPS_COUNT)\
+                or (self.MAX_GROUPS_COUNT is not None and groups_count > self.MAX_GROUPS_COUNT):
+            raise ConstructorGroupsCountError(
+                self.TYPE, len(self.groups), self.MIN_GROUPS_COUNT, self.MAX_GROUPS_COUNT
+            )
+
     @abstractmethod
-    def save(self, rulebase_rule):
-        pass
+    def _check_constructor_params(self):
+        raise NotImplementedError("Subclass should implement this")
+
+    def convert_to_user_constraint_intent(self):
+        """
+        Converts constraint to UserConstraintIntent object.
+
+        For Teacher and Config use while saving constraint into Whylog knowledge base.
+        """
+        return UserConstraintIntent(self.TYPE, self.groups, self.params)
+
+    @classmethod
+    def get_param_names(cls):
+        """
+        Returns names of constraint additional params.
+
+        For Front to display param names to user and then ask user for param contents.
+        """
+        return cls.PARAMS
+
+    @classmethod
+    def get_groups_count(cls):
+        """
+        Returns minimal and maximal count of groups needed to create constraint.
+
+        For Front to ask user for proper count of groups.
+        2, None - at least 2 groups
+        2, 2 - exactly 2 groups
+        """
+        return cls.MIN_GROUPS_COUNT, cls.MAX_GROUPS_COUNT
+
+    @classmethod
+    def verify(cls, group_contents, param_dict):
+        """
+        Verifies constraint for given params in param_dict and groups contents.
+
+        :param param_dict: dict of additional params of constraint
+        :param groups: list of groups contents,
+
+        For LogReader and Teacher verification.
+        It must be optimized as well as possible (for LogReader).
+        """
+
+        raise NotImplementedError("Subclass should implement this")
 
 
 class TimeConstraint(AbstractConstraint):
-    def __init__(self, line_earlier, line_later, min_delta=None, max_delta=None):
-        pass
+    """
+    Time delta between two dates must be greater than 'max_delta' and lower than 'min_delta'
+    """
+
+    TYPE = ConstraintType.TIME_DELTA
+
+    MIN_GROUPS_COUNT = 2
+    MAX_GROUPS_COUNT = 2
+
+    MIN_DELTA = 'min_delta'
+    MAX_DELTA = 'max_delta'
+
+    PARAMS = [MIN_DELTA, MAX_DELTA]
+
+    def __init__(self, groups, param_dict):
+        """
+        :param groups: First element of groups is a group with earlier date.
+
+        I.e:
+        TimeConstraint(
+            [(0, 1), (2, 1)]
+            {'min_delta': datetime.timedelta(seconds=1), 'max_delta': datetime.timedelta(seconds=10)},
+        )
+        """
+        super(TimeConstraint, self).__init__(groups, param_dict)
 
+    def _check_constructor_params(self):
```

Comment:
`if self.PARAMS != keys:` - it is not what I need, because there could be optional parameters. For example, here one of two parameters can be optional. `MIN_DELTA` without `MAX_DELTA` is ok and vice versa.